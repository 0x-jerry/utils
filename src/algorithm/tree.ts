import { toArray } from '../core'
import { isIterable } from '../is'
import { Arrayable, ExtractObjectKeys, Optional } from '../types'

/**
 *
 * DFS (Deep-first search)
 *
 * @param node
 * @param cb
 * @param key default is `'children'`
 */
export function traverseTree<
  T extends {},
  Key extends ExtractObjectKeys<T, Optional<Iterable<any>>, Optional<string>>
>(node: Arrayable<T>, cb: (item: T) => void, key?: Key) {
  const nodes = toArray(node)

  key ??= 'children' as Key

  for (const node of nodes) {
    cb(node)
    const children = node[key]

    if (!isIterable(children)) {
      continue
    }

    traverseTree(children as Arrayable<T>, cb, key)
  }
}

/**
 * @deprecated use `traverseTree` insteadof
 */
export const walkTree = traverseTree
