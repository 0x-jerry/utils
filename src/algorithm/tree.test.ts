import { filterTreeNodes, findNodeInTree, traverse } from './tree.js'

interface TNode {
  num: number
  children?: TNode[]

  // --- type test
  d?: string
  s?: string
  c?: string[]
  e?: string[]
}

describe('tree#traverse', () => {
  const rootNode: TNode = {
    num: 0,
    children: [
      {
        num: 1,
        children: [
          {
            num: 4,
          },
        ],
      },
      {
        num: 2,
      },
    ],
  }

  it('should traverse all tree node in DFS order', () => {
    const nums: number[] = []

    traverse(rootNode, (o) => {
      nums.push(o.num)
    })

    expect(nums).eql([0, 1, 4, 2])
  })

  it('should traverse all tree node in parent node', () => {
    const nums: number[] = []

    traverse(rootNode, (_, parent) => {
      parent && nums.push(parent.num)
    })

    expect(nums).eql([0, 1, 0])
  })

  it('should stop when callback return true', () => {
    let count = 0
    traverse(rootNode, (node) => {
      count++
      return node.num === 4
    })

    expect(count).eql(3)

    // should not stop
    count = 0
    traverse(rootNode, () => {
      count++
    })

    expect(count).eql(4)
  })
})

describe('tree#findNodeInTree', () => {
  const testNode: TNode = {
    num: 4,
  }

  const rootNode: TNode = {
    num: 0,
    children: [
      {
        num: 1,
        children: [testNode],
      },
      {
        num: 2,
      },
    ],
  }

  it('should find node by num === 4', () => {
    let count = 0
    const findedNode = findNodeInTree(rootNode, (node) => {
      count++
      return node.num === 4
    })

    expect(count).eql(3)
    expect(findedNode).toBe(testNode)
  })

  it('should can not find node by num === -2', () => {
    let count = 0
    const findedNode = findNodeInTree(rootNode, (node) => {
      count++
      return node.num === -2
    })

    expect(count).eql(4)
    expect(findedNode).toBeUndefined()
  })
})

describe('tree#filterTreeNodes', () => {
  const rootNode: TNode = {
    num: 0,
    children: [
      {
        num: 1,
        children: [
          {
            num: 4,
          },
        ],
      },
      {
        num: 2,
        children: [
          {
            num: -1,
          },
        ],
      },
    ],
  }

  it('should filter nodes by num >= 2', () => {
    let count = 0
    const newTree = filterTreeNodes(rootNode, (node) => {
      count++
      return node.num >= 2
    })

    expect(count).eql(5)

    expect(newTree).matchSnapshot()
  })

  it('should filter nodes by num < 0', () => {
    let count = 0
    const newTree = filterTreeNodes(rootNode, (node) => {
      count++
      return node.num < 0
    })

    expect(count).eql(5)

    expect(newTree).matchSnapshot()
  })
  it('should filter nodes by num < -2', () => {
    let count = 0
    const newTree = filterTreeNodes(rootNode, (node) => {
      count++
      return node.num < -2
    })

    expect(count).eql(5)

    expect(newTree.length).toBe(0)
  })
})
