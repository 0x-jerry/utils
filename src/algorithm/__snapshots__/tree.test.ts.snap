// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tree#filterTreeNodes > should filter nodes by num < 0 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "num": -1,
          },
        ],
        "num": 2,
      },
    ],
    "num": 0,
  },
]
`;

exports[`tree#filterTreeNodes > should filter nodes by num >= 2 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "num": 4,
          },
        ],
        "num": 1,
      },
      {
        "children": [],
        "num": 2,
      },
    ],
    "num": 0,
  },
]
`;
